refactor: make contract abstract over Borrow

prestwich opened the pull request:
## Motivation

Contracts currently hold `Arc<M>` and have bound `where M: Middleware` . This requires managing arcs across a codebase, which imposes unpleasant requirements on binaries

## Solution

- Modify the contract object to be abstract over `B: Borrow<M>` instead of holding an `Arc<M>`. This allows contracts to be used directly with Cows, non-ARCed middleware, Boxes, other smart pointers, and anything else that uses the `Borrow`  trait.
- rename the modified object `ContractInternal` and create a type alias under the old name, for backwards compatibility

Example use case: 
- you can now use a contract directly with a `Lazy<Provider<Http>>` without having to have an arc in the lazy first

## Inference Limitation

While `FunctionCall` and `ContractInstance` may be used with any type is `Borrow<M>`, there are sometimes multiple choices of borrowed type that satisfy the `M: Middleware` constraint. 

because `Middleware` has an `auto_impl(&, Box, Arc)`, references, boxes, and arcs have multiple `Borrow` options that satisfy the constraint. 
E.g. `Arc<M>` has these impls:
- `impl Borrow<Arc<M>> for Arc<M>`
- `impl Borrow<M> for Arc<M>`
- `impl Borrow<&M> for Arc<M>`

as a result, type inference fails when calling `ContractInstance::new(address, arc_mware)` users must instead specify `ContractInstance::<_, ConcreteMiddlewareType>(address, arc_mware)`

Removing the `auto_impl` will solve this. However, we don't know the extent of reliance on that auto_impl. As a result, using the backwards-compatibility aliases `Contract` and `ContractCall` is still recommended. the Followup work section describes a migration route to fix inference and deprecate the backwards-compatibility aliases

## Followup work

- Change the abigen derived Contract objects to use the generic `ContractInternal` 
- in 2.0.0, eliminate the type alias, and have the  generic version use the `Contract` name

Lib-wide refactoring plan:
1. list further abstraction targets
    - EthEvent
    - ContractDeployer
    - ContractFactory
    - Middlewares (should the stack be `MyMiddleware { inner: Borrow<M> } ` ?)
    - ProviderOracle
    - abigen! output code
3. refactor to use borrow
4. provide backwards compatible type aliases like `type Contract = ContractInstance<Arc<M>, M>`
5. remove `auto_impl(&, Box, Arc)` on `Middleware`
6. write migration guide
7. deprecate backwards-compatibility aliases
8. remove aliases

## PR Checklist

-   [x] Added Tests
-   [x] Added Documentation
-   [x] Updated the changelog
-   [x] Breaking changes

prestwich commented on ethers-contract/src/lib.rs:
```diff
@@ -5,6 +5,9 @@
 pub use contract::Contract;
 
+// pub mod call2;
+// pub mod contract2;
```
artifact of development. need to delete.

prestwich commented on ethers-contract/src/contract.rs:
```diff
@@ -150,32 +150,45 @@ use ethers_core::types::TransactionRequest;
 }
 
-impl<M> std::ops::Deref for Contract<M> {
+pub type Contract<M> = ContractInternal<std::sync::Arc<M>, M>;
```
type aliases ensure backwards compatibility

mattsse reviewed (commented):
this makes it easier indeed, 

especially for http provider, which actually could be made `Clone`

prestwich commented on ethers-contract/src/contract.rs:
```diff
@@ -291,12 +336,13 @@ impl<M: Middleware> Contract<M> {
+        Ok(ContractCallInternal {
             tx,
-            client: Arc::clone(&self.client), // cheap clone behind the Arc
+            client: self.client.clone(),
```
we could make `ContractCallInternal` borrow our B here, by introducing a lifetime to it

I am concerned about the practicality of the `B: Clone` bound, for borrow types _without_ clone. are there commonly used borrow types that are _NOT_ clone? 🤔

DaniPopes replied:
It can be the case when `B: Borrow<M>` is equal to `M` and `M` is not Clone, since `Borrow` has a blanket implemention for any `T: ?Sized`. Otherwise it would mostly be the standard smart pointers. So I'd say if all our `Middleware` implement `Clone` this is fine

prestwich replied:
unfortunately, not all of our middleware implement Clone, IIRC. e.g. I think `SignerMiddleware<LedgerSigner, _>` does not 🤔

DaniPopes replied:
Ah yeah right, I guess this will need to be specified in a doc somewhere that it is recommended wrapping `M` with a smart pointer or using references to avoid cloning data or for cases where that it is actually not possible

prestwich replied:
I'm going to make a couple toy examples for when `B` is `M` or `&M`. (`&M` is `Clone` however, so that should be fine)

prestwich replied:
![Screenshot 2023-01-27 at 12 19 12 PM](https://user-images.githubusercontent.com/10149425/215152256-898d606d-ad3e-4d52-ba22-3625a45feab8.png)

prestwich replied:
I've also confirmed that `NonceManagerMiddleware` is not Clone

prestwich replied:
so the place where I user would run into this is:

- instantiating a contract with an _owned_ **non-clone** middleware

prestwich replied:
slight adjustment to screenshot. 3rd example is wrong, as M is `&`. corrected version here:

![Screenshot 2023-01-27 at 12 27 58 PM](https://user-images.githubusercontent.com/10149425/215154007-cbcafdce-6390-480e-8bbb-c010eb1751cd.png)

gakonst replied:
Is that a big problem? Can't we pass these Middlewares by reference, instead of owned? And otherwise, just Arc them?

prestwich replied:
> Is that a big problem? Can't we pass these Middlewares by reference, instead of owned? And otherwise, just Arc them?

it's a not-immediately-obivous issue that users will run into and ask about. It compounds with the ambiguity caused by `auto_imp(&, Box, Arc)` to make this more-powerful API extremely difficult to use

if you call `ContractInstance::new(address, &provider)` you get a type ambiguity around `M`. is `M` `Provider` or `&Provider`. Both implement `Middleware` and both can be borrowed from `&Provider`. Same ambiguity for `Arc`

The follow up work (will copy this to original post) to get this usable by beginner rustaceans is something like:
```
1. list further abstraction targets
2. refactor to use borrow
3. provide backwards compatible type aliases like `type Contract = ContractInstance<Arc<M>, M>`
4. remove `auto_impl(&, Box, Arc)` on `Middleware`
5. write migration guide
6. deprecate backwards-compatibility aliases
7. remove aliases
```

prestwich commented on ethers-contract/src/base.rs:
```diff
@@ -200,7 +200,7 @@ impl BaseContract {
         client: impl Into<Arc<M>>,
     ) -> Contract<M> {
-        Contract::new(address, self, client)
+        Contract::new(address, self, client.into())
```
should this be generic over B? I think so. I'll change it in a minute

mattsse commented on ethers-contract/src/contract.rs:
```diff
@@ -209,14 +229,14 @@ impl<M: Middleware> Contract<M> {
-        client: impl Into<Arc<M>>,
-    ) -> Self {
-        Self { base_contract: abi.into(), client: client.into(), address: address.into() }
+    pub fn new(address: impl Into<Address>, abi: impl Into<BaseContract>, client: B) -> Self {
```
this change makes more sense, 
because `impl Into Arc` is kinda dumb, especially if you need already have an `Arc<P>`

prestwich commented:
currently checking out the test breakage. looks like this will affect type inference

prestwich commented:
```
error[E0283]: type annotations needed
   --> ethers-contract/tests/it/contract.rs:431:14
    |
431 |             .connect(client2.clone())
    |              ^^^^^^^ cannot infer type of the type parameter `N` declared on the associated function `connect`
    |
    = note: multiple `impl`s satisfying `Arc<ethers_providers::Provider<Http>>: Borrow<_>` found in the following crates: `alloc`, `core`:
            - impl<T> Borrow<T> for Arc<T>
              where T: ?Sized;
            - impl<T> Borrow<T> for T
              where T: ?Sized;
note: required by a bound in `ethers_contract::contract::ContractInternal::<B, M>::connect`
   --> /Users/james/devel/ethers-rs/ethers-contract/src/contract.rs:269:12
    |
269 |         C: Borrow<N>,
    |            ^^^^^^^^^ required by this bound in `ethers_contract::contract::ContractInternal::<B, M>::connect`
help: consider specifying the type arguments in the function call
    |
431 |             .connect::<C, N>(client2.clone())
    |                     ++++++++
```

the test failures are all driven by failed inference on `connect`. because `connect` is now generic over `B`, and `B: Borrow<M>` and `M: Borrow<M>`, both B and M are valid args for `connect()`. the compiler doesn't know whether it should be using the `B` or borrowing `M` from it and using `M`

this is only an issue when _switching_ the middleware connected to a contract, and the solution is to write out `connect::<Arc<_>, _>()`. I think this is acceptable (although it is a breaking change)

prestwich commented:
workaround:
- changed params to `connect` to use an `arc` to leave existing code un-broken
- added `connect_with` with the borrow api, which requires more type ascription

prestwich commented:
anyone have a better name than `ContractInternal` ? 🤔

mattsse commented on ethers-contract/src/call.rs:
```diff
@@ -73,33 +72,45 @@ pub enum ContractError<M: Middleware> {
 #[must_use = "contract calls do nothing unless you `send` or `call` them"]
 /// Helper for managing a transaction before submitting it to a node
-pub struct ContractCall<M, D> {
+pub struct ContractCallInternal<B, M, D> {
```
What is the motivation for this renaming?
so that we can use `ContractCall` for backwards compat?

The "internal" postfix somewhat irritates me as well.
I guess, technically this is a `ContractFunctionCall` or `FunctionCall`?

prestwich replied:
`FunctionCall` seems like a good name, as contract calls may not have functions (e.g. fallback or receive invocations)

mattsse commented on ethers-contract/src/contract.rs:
```diff
@@ -145,37 +145,73 @@ use ethers_core::types::TransactionRequest;
 }
 
-impl<M> std::ops::Deref for Contract<M> {
+/// `Contract` is a [`ContractInternal`] object with an `Arc` middleware.
+/// This type alias exists to preserver backwards compatibility with
+/// non-abstract Contracts.
+///
+/// For full usage docs, see [`ContractInternal`]
+pub type Contract<M> = ContractInternal<std::sync::Arc<M>, M>;
```
nit: type aliases at the beginning of the file/ above struct def, as for `ContractCall`

mattsse commented on ethers-contract/src/contract.rs:
```diff
@@ -145,37 +145,73 @@ use ethers_core::types::TransactionRequest;
 }
 
-impl<M> std::ops::Deref for Contract<M> {
+/// `Contract` is a [`ContractInternal`] object with an `Arc` middleware.
+/// This type alias exists to preserver backwards compatibility with
+/// non-abstract Contracts.
+///
+/// For full usage docs, see [`ContractInternal`]
+pub type Contract<M> = ContractInternal<std::sync::Arc<M>, M>;
```
should we perhaps mark the aliases as deprecated?

prestwich replied:
adding a deprecated tag results in close to 200 warnings in the ethers codebase. To avoid noise in the PR, I am adding in a commented-out deprecated attribute. I'll uncomment it and change all of our internal usage as a last step.

prestwich replied:
update:

see top comment for future plans wrt this. Aliases will not be deprecated in this PR. We will deprecate after we have a straightforward and beginner-friendly migration plan

mattsse commented on ethers-contract/src/contract.rs:
```diff
@@ -145,37 +145,73 @@ use ethers_core::types::TransactionRequest;
 /// [`method`]: method@crate::Contract::method
 #[derive(Debug)]
-pub struct Contract<M> {
+pub struct ContractInternal<B, M> {
```
The same concerns re naming here.
`Internal` is likely confusing, but for this one, I don't have any suggestions at this point.

hmm, compared to `BaseContract` this type represents an `Existing`, `Deployed` contract since it has an `address`, so perhaps we include this in the name?

prestwich replied:
Naming for `Contract` is inherited from ethers.js, and implies an interface with an address and a provider. I believe `BaseContract` was backfilled to make sense given `Contract` has a connection to the chain.

proposed better nomenclature:

- `Interface` -- currently called `BaseContract` provides encoding/decoding functionality
   - alias to `BaseContract` for backwards compat
   - maybe de-duplicate the hashmap via a global cache so we stop doing a bunch of allocs on clone?
- `ContractInstance` -- currently `ContractInternal`. Wraps `BaseContract` with an address and a connection
  - alias version with arc to `Contract` for backwards compat

mattsse reviewed (changes requested):
I like this change, 
using `Borrow` should allow us to create a `Contract` that uses a reference to a provider (`&M`) right? do we have an example for this.

so it's possible to 
1. connect ws
2. create and consume `Contract` with `&ws`
3. return ownership of `ws`

so "Arc<Ws>" should no longer be mandatory for every "Contract" interaction, right?

gakonst commented on ethers-contract/src/contract.rs:
```diff
@@ -240,26 +289,50 @@ impl<M: Middleware> Contract<M> {
+    /// Returns a new contract instance using the provided client
+    ///
+    /// Clones `self` internally
+    pub fn connect<N>(&self, client: Arc<N>) -> ContractInstance<Arc<N>, N>
```
Should this also be `#[must_use]`?

prestwich commented:
@gakonst @mattsse @DaniPopes 

Please see updated notes at top of comment thread. I believe this is ready for final review and merge 🎉

gakonst commented on ethers-contract/src/call.rs:
```diff
@@ -73,33 +72,49 @@ pub enum ContractError<M: Middleware> {
     ContractNotDeployed,
 }
 
+/// `ContractCall` is a [`FunctionCall`] object with an [`Arc`] middleware.
```
```suggestion
/// `ContractCall` is a [`FunctionCall`] object with an [`std::sync::Arc`] middleware.
```

gakonst reviewed (approved):
Supportive - pending @mattsse

mattsse reviewed (approved):
I think this is step in the right direction.

I guess there could be some quirks wrt Arc and its auto_impls but we can figure things out when we encounter them.
